platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -434,41 +435,59 @@ public void endWriting() throws IOException {
-                    continue;
-                }
-                if (res.error != null) {
-                    throw new RuntimeException(res.error);
+            try {
+                //  write to files
+                fileCollection.startWriting();
+                int processed = 0;
+                while (processed < size) {
+                    final ReadBucketResult<K> res = queue.poll();
+                    if (res == null) {
+                        continue;
+                    }
+                    if (res.error != null) {
+                        throw new RuntimeException(res.error);
+                    }
+                    try (final Bucket<K> bucket = res.bucket) {
+                        final int bucketIndex = bucket.getBucketIndex();
+                        if (bucket.getBucketEntryCount() == 0) {
+                            // bucket is missing or empty, remove it from the index
+                            bucketIndexToBucketLocation.remove(bucketIndex);
+                        } else {
+                            // save bucket
+                            final long bucketLocation = fileCollection.storeDataItem(bucket);
+                            // update bucketIndexToBucketLocation
+                            bucketIndexToBucketLocation.put(bucketIndex, bucketLocation);
+                        }
+                    } finally {
+                        processed++;
+                    }
                 }
-                try (final Bucket<K> bucket = res.bucket) {
-                    final int bucketIndex = bucket.getBucketIndex();
-                    if (bucket.getBucketEntryCount() == 0) {
-                        // bucket is missing or empty, remove it from the index
-                        bucketIndexToBucketLocation.remove(bucketIndex);
-                    } else {
-                        // save bucket
-                        final long bucketLocation = fileCollection.storeDataItem(bucket);
-                        // update bucketIndexToBucketLocation
-                        bucketIndexToBucketLocation.put(bucketIndex, bucketLocation);
+                // close files session
+                final DataFileReader<Bucket<K>> dataFileReader = fileCollection.endWriting(0, numOfBuckets);
+                // we have updated all indexes so the data file can now be included in merges
+                dataFileReader.setFileCompleted();
+            } finally {
+                endWritingRunning.set(false);
+                // When the flushing thread is interrupted, the chances are all flushExecutor threads
+                // are stuck waiting on reusable buckets to become available. To unblock them, process
+                // remaining results in the queue and release all updated buckets
+                ReadBucketResult<K> res;
+                while ((res = queue.poll()) != null) {
+                    if (res.bucket != null) {
+                        res.bucket.close();
```

Comment:
A possibility of a bucket leak still remains if there is an exception: a task may see endWritingRunning as true, start working and then add its result after this loop finishes.

Reply:
This is correct, but it should not be a problem. If the flushing thread is interrupted or experienced an exception, it means the data source (and HalfDiskHashMap) needs to be recreated anyway. This is what happens, when reconnect fails. The goal here is to unblock all threads in `flushExecutor`, even if some buckets are read, updated, but never flushed to disk.